src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs

```diff
@@ -311,5 +345,32 @@ private Uri CreateFullEndpointUri(Uri endPoint)
                 }
             }
         }
+
+        private IDictionary<string, string> ParseChallengeData(HttpResponse response)
+        {
+            IDictionary<string, string> data = new Dictionary<string, string>();
+            string wwwAuthenticate = response.Headers.GetValues(WwwAuthenticateHeader).FirstOrDefault();
+            wwwAuthenticate = wwwAuthenticate.Substring(PKeyAuthName.Length + 1);
+            List<string> headerPairs = CoreHelpers.SplitWithQuotes(wwwAuthenticate, ',');
+            foreach (string pair in headerPairs)
+            {
+                List<string> keyValue = CoreHelpers.SplitWithQuotes(pair, '=');
+                data.Add(keyValue[0].Trim(), keyValue[1].Trim().Replace("\"", ""));
+            }
+
+            return data;
+        }
+
+        private bool IsDeviceAuthChallenge(HttpResponse response, bool respondToDeviceAuthChallenge)
+        {
+            return _deviceAuthManager.CanHandleDeviceAuthChallenge
+                   && response != null
+                   && respondToDeviceAuthChallenge
+                   && response.Headers != null
+                   && response.StatusCode == HttpStatusCode.Unauthorized
+                   && response.Headers.Contains(WwwAuthenticateHeader)
+                   && response.Headers.GetValues(WwwAuthenticateHeader).FirstOrDefault()
```

Comment:
I would use .Any() instead of .First() in the unlikely event that the server returns 2 values there.